host.py

```diff
@@ -292,20 +293,45 @@ def copy_to(self, src_file: str, dst_file: str) -> None:
     def need_sudo(self) -> None:
         self.sudo_needed = True
 
-    def run(self, cmd: str, log_level: int = logging.DEBUG, env: dict[str, str] = os.environ.copy()) -> Result:
+    def run(
+        self,
+        cmd: str,
+        log_level: int = logging.DEBUG,
+        env: Optional[dict[str, str]] = None,
+        *,
+        env_extra: Optional[Mapping[str, Optional[str]]] = None,
+    ) -> Result:
```

Comment:
This style is bad. All the args should be on one line. If the line is too long, the code needs to be refactored. It's a red flag.

Reply:
The line gets longer, because arguments are added. They are added for a reason. It's not only this one, follow-up patches add other arguments  for different reasons.

How could this be solved instead?


---

Btw, this style suggestion is not agreeable. Line breaks are fine, an so thinks `python-black` and PEP8 which add them eagerly. Also, a larger number of arguments can in some(!!) cases be perfectly adequate solutions. See `subprocess.run()`. There is nothing to refactor there. This is thanks to keyword(only)-arguments and default-values in Python. In C that would be different.

Reply:
> The line gets longer, because arguments are added. They are added for a reason...

Well, I'm saying: "long lines are a red flag" which then implies "many args are a red flag".

First: `cmd: str | Iterable[str]`. Why can't it be a simple `str`? Why are we adding `| Iterable[str]`? 
Second: `env_extra: Optional[Mapping[str, Optional[str]]] = None,` This I still don't like since it's confusing given that we also have `env`.
Third: `cwd: Optional[str] = None`, it probably makes sense to move that to the constructor and add a cd argument to the `Host` object.

The following line is 102 characters long:
`def run(self, cmd: str, log_level: int = logging.DEBUG, env: Optional[dict[str, str]] = None) -> Result`

I understand partially that you are saying "they are there for a reason". Sure, I can write whole CDA as 1 huge function that takes in 1000 arguments, but is that readable/usable/maintainable?

Please think more about the code and push a new version.